File: ios/AudioPlayerEngine.swift
import AVFoundation
import Foundation
import UIKit

/// Thin wrapper around `AVPlayer` that exposes the events the rest of the
/// component needs: load lifecycle, periodic time updates, end-of-track,
/// and rate / seek control. All callbacks fire on the main thread.
final class AudioPlayerEngine {

    // MARK: - State

    enum State {
        case idle
        case loading
        case ready
        case ended
        case error
    }

    private(set) var state: State = .idle {
        didSet {
            guard oldValue != state else { return }
            // If the user tapped "play" while we were still loading, apply
            // playback synchronously *before* firing onStateChange so the
            // single notification reflects the final state (state == .ready
            // AND isPlaying == true). Without this we'd fire one event with
            // isPlaying=false (causing a brief play-icon flash) then a
            // second one once startInternal flips it.
            if state == .ready, pendingStart {
                pendingStart = false
                startPlaybackInternal()
            }
            onStateChange?()
        }
    }

    /// "Play once ready" intent recorded during the `.loading` state.
    /// Cleared on pause / reset / source change.
    private var pendingStart: Bool = false

    private(set) var durationMs: Int = 0
    private(set) var currentMs: Int = 0
    /// `true` while the engine is in a "playing" state from the user's POV
    /// (i.e. `play()` was called and we haven't been paused or ended).
    /// We track this explicitly because `AVPlayer.timeControlStatus` flips
    /// between `playing` / `waitingToPlayAtSpecifiedRate` during buffering.
    private(set) var isPlaying: Bool = false
    private(set) var rate: Float = 1.0
    var loop: Bool = false

    // MARK: - Callbacks

    var onLoad: ((Int) -> Void)?
    var onLoadError: ((String) -> Void)?
    var onStateChange: (() -> Void)?
    var onTimeUpdate: ((Int, Int) -> Void)?
    var onEnded: (() -> Void)?

    // MARK: - Private

    private let player = AVPlayer()
    private var currentItem: AVPlayerItem?
    private var timeObserver: Any?
    private var statusObservation: NSKeyValueObservation?
    private var endObservation: NSObjectProtocol?

    init() {
        player.actionAtItemEnd = .pause
        // Be eager about playback readiness: AVPlayer's default behaviour
        // is to accumulate a generous forward buffer before flipping
        // `.readyToPlay`, which can keep the loading spinner on screen
        // for many seconds even after enough data has arrived to start
        // playback. With this disabled, `.readyToPlay` fires as soon as
        // the item has decodable samples queued up.
        player.automaticallyWaitsToMinimizeStalling = false
        // The library does not configure AVAudioSession by default to avoid
        // surprises; opt-in via `setBackgroundPlaybackEnabled(true)` instead.
    }

    /// Configure the shared `AVAudioSession` so audio keeps playing when the
    /// host app is backgrounded. Requires the host app to have the "Audio,
    /// AirPlay, and Picture in Picture" Background Mode enabled in Info.plist.
    ///
    /// Calling with `false` is a no-op — once the session category has been
    /// switched to `.playback` we leave it alone (the host app may have its
    /// own audio session management we don't want to step on).
    func setBackgroundPlaybackEnabled(_ enabled: Bool) {
        guard enabled else { return }
        let session = AVAudioSession.sharedInstance()
        // Don't churn the session if it's already in a playback-capable mode.
        if session.category == .playback || session.category == .playAndRecord {
            try? session.setActive(true, options: [])
            return
        }
        do {
            try session.setCategory(.playback, mode: .default, options: [])
            try session.setActive(true, options: [])
        } catch {
            // Silently ignore — host app likely manages its own session.
        }
    }

    deinit {
        teardownObservers()
        // ARC will release `player` immediately after this returns, and
        // AVPlayer is supposed to stop on dealloc — but in practice it
        // can survive briefly if a background `AVAudioSession` keeps it
        // alive. Pausing + dropping the current item synchronously here
        // guarantees the audio actually stops, even if our `reset()` was
        // never called.
        player.pause()
        player.replaceCurrentItem(with: nil)
    }

    // MARK: - Public API

    func setSource(url: URL) {
        teardownObservers()

        pendingStart = false
        state = .loading
        currentMs = 0
        durationMs = 0
        isPlaying = false

        // `AVPlayer.rate` is independent of the current item and survives
        // `replaceCurrentItem(with:)` — so if the previous item was playing
        // (rate=1.0) and we don't pause here, the new item auto-resumes the
        // moment its status flips to `.readyToPlay` (because we also set
        // `automaticallyWaitsToMinimizeStalling = false`). That manifests
        // as "audio plays on its own after a mount / unmount / remount cycle,
        // and the play button still shows the play icon" because engine
        // bookkeeping says `isPlaying = false` but AVPlayer is happily
        // playing the new item at the leftover rate.
        player.pause()

        let item = AVPlayerItem(url: url)
        currentItem = item
        statusObservation = item.observe(\.status, options: [.new]) { [weak self] item, _ in
            guard let self = self else { return }
            DispatchQueue.main.async {
                switch item.status {
                case .readyToPlay:
                    let durSeconds = item.duration.seconds
                    if durSeconds.isFinite, durSeconds > 0 {
                        self.durationMs = Int(durSeconds * 1000)
                    } else {
                        self.durationMs = 0
                    }
                    self.state = .ready
                    self.onLoad?(self.durationMs)
                case .failed:
                    let message = item.error?.localizedDescription ?? "Unknown player error"
                    self.state = .error
                    self.onLoadError?(message)
                default:
                    break
                }
            }
        }

        endObservation = NotificationCenter.default.addObserver(
            forName: .AVPlayerItemDidPlayToEndTime,
            object: item,
            queue: .main
        ) { [weak self] _ in
            self?.handleEnd()
        }

        player.replaceCurrentItem(with: item)

        let interval = CMTime(value: 1, timescale: 30)
        timeObserver = player.addPeriodicTimeObserver(
            forInterval: interval,
            queue: .main
        ) { [weak self] time in
            guard let self = self, self.durationMs > 0 else { return }
            let seconds = time.seconds
            if seconds.isFinite, seconds >= 0 {
                self.currentMs = min(self.durationMs, Int(seconds * 1000))
                self.onTimeUpdate?(self.currentMs, self.durationMs)
            }
        }
    }

    func play() {
        if state == .loading {
            // Audio isn't buffered yet — record the intent and let the
            // state setter resume playback the instant we transition to
            // `.ready`. We deliberately don't fire onStateChange here so
            // the play/pause button stays as the loading spinner instead
            // of briefly flipping to a "pause" icon while still loading.
            pendingStart = true
            return
        }
        guard state == .ready || state == .ended else { return }
        // Already running — skip so we don't fire a redundant
        // onStateChange every time `applyControlledState()` is called.
        if isPlaying && state == .ready { return }
        // Clear any stale pending intent before we drive state changes,
        // so the state-setter doesn't try to "resume" again.
        pendingStart = false
        if state == .ended {
            player.seek(to: .zero)
            currentMs = 0
            state = .ready
        }
        startPlaybackInternal()
        onStateChange?()
    }

    func pause() {
        // Cancel any queued "play once ready" intent — the user explicitly
        // wants playback to stay paused.
        pendingStart = false
        guard isPlaying else { return }
        isPlaying = false
        player.pause()
        onStateChange?()
    }

    func toggle() {
        if isPlaying { pause() } else { play() }
    }

    /// Seek to position in milliseconds. Uses an exact-tolerance seek so the
    /// playhead lands on the requested sample even for VBR mp3.
    func seek(toMs ms: Int, completion: (() -> Void)? = nil) {
        let clamped = max(0, min(durationMs, ms))
        currentMs = clamped
        let target = CMTime(value: CMTimeValue(clamped), timescale: 1000)
        player.seek(to: target, toleranceBefore: .zero, toleranceAfter: .zero) { _ in
            completion?()
        }
    }

    func setRate(_ newRate: Float) {
        let clamped = max(0.25, min(4.0, newRate))
        rate = clamped
        if isPlaying {
            player.rate = clamped
        }
    }

    func reset() {
        pendingStart = false
        teardownObservers()
        // Pause BEFORE dropping the current item — otherwise `player.rate`
        // (which AVPlayer treats as orthogonal to the current item) sticks
        // at whatever it was last set to and silently kicks the next item
        // into playback the moment it becomes ready. See `setSource(url:)`
        // for the long version.
        player.pause()
        player.replaceCurrentItem(with: nil)
        currentItem = nil
        isPlaying = false
        currentMs = 0
        durationMs = 0
        rate = 1.0
        state = .idle
    }

    // MARK: - Internal

    /// Common play-start sequence shared by `play()` and the in-line
    /// resume from the `state` setter when transitioning to `.ready`
    /// with a queued tap intent. Does NOT fire `onStateChange` — callers
    /// are responsible for that (so we can batch a single notification).
    private func startPlaybackInternal() {
        isPlaying = true
        player.rate = rate
        // Calling `player.play()` after setting rate keeps the rate sticky
        // even after a previous .pause() reset it to 0.
        player.play()
        player.rate = rate
    }

    private func handleEnd() {
        if loop {
            player.seek(to: .zero)
            currentMs = 0
            if isPlaying {
                player.rate = rate
                player.play()
                player.rate = rate
            }
        } else {
            isPlaying = false
            currentMs = durationMs
            state = .ended
            onTimeUpdate?(currentMs, durationMs)
            onEnded?()
        }
    }

    private func teardownObservers() {
        if let timeObserver = timeObserver {
            player.removeTimeObserver(timeObserver)
        }
        timeObserver = nil
        statusObservation?.invalidate()
        statusObservation = nil
        if let endObservation = endObservation {
            NotificationCenter.default.removeObserver(endObservation)
        }
        endObservation = nil
    }
}

File: android/src/main/java/com/waveformrecorder/AudioPlayerEngine.kt
package com.waveformrecorder

import android.Manifest
import android.content.pm.PackageManager
import android.media.MediaPlayer
import android.media.PlaybackParams
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.PowerManager
import android.util.Log

/**
 * Thin wrapper around the built-in `android.media.MediaPlayer` that exposes
 * the events the rest of the component needs: load lifecycle, periodic time
 * updates, end-of-track, and rate / seek control. All callbacks fire on the
 * main thread.
 *
 * Forked verbatim from the sibling
 * `react-native-waveform-player` library so the recorder can ship a
 * preview state without taking a runtime dependency on the player package.
 */
class AudioPlayerEngine {

    enum class State { IDLE, LOADING, READY, ENDED, ERROR }

    var state: State = State.IDLE
        private set

    var durationMs: Int = 0
        private set

    var currentMs: Int = 0
        private set

    var isPlaying: Boolean = false
        private set

    var rate: Float = 1.0f
        private set

    var loop: Boolean = false

    // Callbacks
    var onLoad: ((Int) -> Unit)? = null
    var onLoadError: ((String) -> Unit)? = null
    var onStateChange: (() -> Unit)? = null
    var onTimeUpdate: ((Int, Int) -> Unit)? = null
    var onEnded: (() -> Unit)? = null

    private var player: MediaPlayer? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private val progressRunnable = object : Runnable {
        override fun run() {
            val p = player ?: return
            if (state != State.READY || !isPlaying) return
            try {
                currentMs = p.currentPosition.coerceIn(0, durationMs)
                onTimeUpdate?.invoke(currentMs, durationMs)
            } catch (_: IllegalStateException) {
                // Player isn't ready / has been released; just stop polling.
            }
            mainHandler.postDelayed(this, 33L)
        }
    }

    /** Cached rate to be applied once the player transitions to Started. */
    private var pendingRate: Float? = null

    /**
     * "Play once ready" intent recorded if `play()` is called while the
     * player is still in `LOADING`. Cleared on pause / reset / source
     * change. Resumed automatically when we transition into `READY`.
     */
    private var pendingStart: Boolean = false

    /**
     * Cached "would like a partial wake lock" flag. Applied each time the
     * underlying `MediaPlayer` is (re-)created. Requires the host app to
     * declare `WAKE_LOCK` in its manifest — without it `setWakeMode` throws
     * `SecurityException`, which we catch and ignore so playback still works
     * (it just won't survive device sleep with the screen off).
     */
    private var wakeModeEnabled: Boolean = false

    fun setSource(context: android.content.Context, uri: String) {
        reset()
        setStateInternal(State.LOADING)
        try {
            val mp = MediaPlayer()
            applyWakeMode(context, mp)
            mp.setOnPreparedListener { prepared ->
                durationMs = try { prepared.duration.coerceAtLeast(0) } catch (_: Exception) { 0 }
                setStateInternal(State.READY)
                onLoad?.invoke(durationMs)
            }
            mp.setOnCompletionListener {
                if (loop) {
                    try {
                        it.seekTo(0)
                        currentMs = 0
                        if (isPlaying) {
                            it.start()
                            applyPendingRate()
                        }
                    } catch (_: Exception) {}
                } else {
                    isPlaying = false
                    currentMs = durationMs
                    setStateInternal(State.ENDED)
                    onTimeUpdate?.invoke(currentMs, durationMs)
                    onEnded?.invoke()
                    stopProgressLoop()
                }
            }
            mp.setOnErrorListener { _, what, extra ->
                isPlaying = false
                stopProgressLoop()
                setStateInternal(State.ERROR)
                onLoadError?.invoke("MediaPlayer error: what=$what extra=$extra")
                true
            }
            try {
                if (uri.startsWith("http://") || uri.startsWith("https://")) {
                    mp.setDataSource(uri)
                } else {
                    mp.setDataSource(context, Uri.parse(uri))
                }
            } catch (e: Exception) {
                Log.e(TAG, "setDataSource failed", e)
                setStateInternal(State.ERROR)
                onLoadError?.invoke("setDataSource failed: ${e.message}")
                mp.release()
                return
            }
            try {
                mp.prepareAsync()
            } catch (e: Exception) {
                Log.e(TAG, "prepareAsync failed", e)
                setStateInternal(State.ERROR)
                onLoadError?.invoke("prepareAsync failed: ${e.message}")
                mp.release()
                return
            }
            player = mp
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create MediaPlayer", e)
            setStateInternal(State.ERROR)
            onLoadError?.invoke(e.message ?: "Unknown error")
        }
    }

    fun play() {
        if (state == State.LOADING) {
            pendingStart = true
            return
        }
        val p = player ?: return
        if (state != State.READY && state != State.ENDED) return
        if (isPlaying && state == State.READY) return
        try {
            pendingStart = false
            if (state == State.ENDED) {
                p.seekTo(0)
                currentMs = 0
                setStateInternal(State.READY)
            }
            p.start()
            isPlaying = true
            applyPendingRate()
            startProgressLoop()
            onStateChange?.invoke()
        } catch (e: Exception) {
            isPlaying = false
            stopProgressLoop()
            Log.e(TAG, "play failed", e)
            onStateChange?.invoke()
        }
    }

    fun pause() {
        pendingStart = false
        val p = player ?: return
        if (!isPlaying) return
        try {
            p.pause()
            isPlaying = false
            stopProgressLoop()
            onStateChange?.invoke()
        } catch (e: Exception) {
            Log.e(TAG, "pause failed", e)
        }
    }

    fun toggle() {
        if (isPlaying) pause() else play()
    }

    fun seekToMs(ms: Int) {
        val p = player ?: return
        val clamped = ms.coerceIn(0, durationMs.coerceAtLeast(0))
        currentMs = clamped
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                p.seekTo(clamped.toLong(), MediaPlayer.SEEK_CLOSEST)
            } else {
                p.seekTo(clamped)
            }
        } catch (e: Exception) {
            Log.e(TAG, "seekTo failed", e)
        }
    }

    fun setRate(newRate: Float) {
        val clamped = newRate.coerceIn(0.25f, 4.0f)
        rate = clamped
        val p = player ?: return
        if (isPlaying) {
            try {
                p.playbackParams = PlaybackParams().setSpeed(clamped)
                pendingRate = null
            } catch (e: Exception) {
                Log.e(TAG, "setPlaybackParams failed", e)
                pendingRate = clamped
            }
        } else {
            pendingRate = clamped
        }
    }

    /**
     * Toggle whether `MediaPlayer.setWakeMode(PARTIAL_WAKE_LOCK)` should be
     * applied. Takes effect on the next `setSource(...)` call (and the
     * current player, if any).
     */
    fun setBackgroundPlaybackEnabled(context: android.content.Context, enabled: Boolean) {
        wakeModeEnabled = enabled
        player?.let { applyWakeMode(context, it) }
    }

    private fun applyWakeMode(context: android.content.Context, mp: MediaPlayer) {
        if (!wakeModeEnabled) return
        val granted = context.checkSelfPermission(Manifest.permission.WAKE_LOCK) ==
            PackageManager.PERMISSION_GRANTED
        if (!granted) {
            Log.w(
                TAG,
                "playInBackground=true but WAKE_LOCK permission is not granted — skipping " +
                    "MediaPlayer.setWakeMode. Background playback still works while the screen " +
                    "is on. To survive device sleep, add " +
                    "`<uses-permission android:name=\"android.permission.WAKE_LOCK\"/>` " +
                    "to your app manifest."
            )
            return
        }
        try {
            mp.setWakeMode(context.applicationContext, PowerManager.PARTIAL_WAKE_LOCK)
        } catch (e: Exception) {
            Log.w(TAG, "setWakeMode failed", e)
        }
    }

    private fun applyPendingRate() {
        val p = player ?: return
        val target = pendingRate ?: rate
        try {
            p.playbackParams = PlaybackParams().setSpeed(target)
            pendingRate = null
        } catch (e: Exception) {
            Log.e(TAG, "applyPendingRate failed", e)
        }
    }

    fun reset() {
        stopProgressLoop()
        try { player?.reset() } catch (_: Exception) {}
        try { player?.release() } catch (_: Exception) {}
        player = null
        isPlaying = false
        currentMs = 0
        durationMs = 0
        rate = 1.0f
        pendingRate = null
        pendingStart = false
        setStateInternal(State.IDLE)
    }

    private fun setStateInternal(newState: State) {
        if (state == newState) return
        state = newState
        if (newState == State.READY && pendingStart) {
            pendingStart = false
            tryStartPlaybackInternal()
        }
        onStateChange?.invoke()
    }

    private fun tryStartPlaybackInternal() {
        val p = player ?: return
        try {
            p.start()
            isPlaying = true
            applyPendingRate()
            startProgressLoop()
        } catch (e: Exception) {
            isPlaying = false
            stopProgressLoop()
            Log.e(TAG, "deferred play failed", e)
        }
    }

    private fun startProgressLoop() {
        stopProgressLoop()
        mainHandler.post(progressRunnable)
    }

    private fun stopProgressLoop() {
        mainHandler.removeCallbacks(progressRunnable)
    }

    companion object {
        private const val TAG = "AudioPlayerEngine"
    }
}
